### PR TITLE
Add certificates encryption

### PR DESCRIPTION
**Pull request by corest**
Towards https://github.com/giantswarm/giantswarm/issues/2816

### CODE REVIEW

File: service/controller/v5/cloudconfig/render.go [hunk 1]
```diff
@@ -32,12 +32,16 @@ func renderCalicoAzureFile(params calicoAzureFileParams) (k8scloudconfig.FileAss
 	return file, nil
 }
 
-func renderCertificatesFiles(certFiles certs.Files) ([]k8scloudconfig.FileAsset, error) {
+func renderCertificatesFiles(encrypter Encrypter, certFiles certs.Files) ([]k8scloudconfig.FileAsset, error) {
 	var certsMeta []k8scloudconfig.FileMetadata
 	for _, f := range certFiles {
+		encryptedData, err := encrypter.EncryptCFBBase64(f.Data)
+		if err != nil {
+			return []k8scloudconfig.FileAsset{}, microerror.Mask(err)
```
Comment (xh3b4sd) [inline service/controller/v5/cloudconfig/render.go]:
```suggestion
			return nil, microerror.Mask(err)
```

File: service/controller/v5/resource/instance/types.go [hunk 2]
```diff
@@ -38,5 +40,8 @@ func newNodeOSImageCoreOS() nodeOSImage {
 // SmallCloudconfigConfig represents the data structure required for executing
 // the small cloudconfig template.
 type SmallCloudconfigConfig struct {
-	BlobURL string
+	BlobURL       string
```
Comment (xh3b4sd) [inline service/controller/v5/resource/instance/types.go]: For me again, why is the blob URL necessary here?

Comment (corest) [reply service/controller/v5/resource/instance/types.go]: There is a limit for the size of the machine scaleset userdata field. We can't put full ignition there. So we put small ignition, which then appends main ignition config from url. This small ignition also includes an encryption key for certificates.

Comment (corest) [reply service/controller/v5/resource/instance/types.go]: It is the same approach as in aws-operator https://github.com/giantswarm/aws-operator/blob/master/service/controller/v23/templates/cloudconfig/small.go#L9

**Comment by corest** [comment]
Stuff works. @giantswarm/sig-operator ptal

File: service/controller/v5/cloudconfig/render.go [hunk 3]
```diff
@@ -32,12 +33,17 @@ func renderCalicoAzureFile(params calicoAzureFileParams) (k8scloudconfig.FileAss
 	return file, nil
 }
 
-func renderCertificatesFiles(certFiles certs.Files) ([]k8scloudconfig.FileAsset, error) {
+func renderCertificatesFiles(encrypter encrypter.Encrypter, certFiles certs.Files) ([]k8scloudconfig.FileAsset, error) {
 	var certsMeta []k8scloudconfig.FileMetadata
 	for _, f := range certFiles {
+		encryptedData, err := encrypter.EncryptCFB(f.Data)
+
```
Comment (xh3b4sd) [inline service/controller/v5/cloudconfig/render.go]:
```suggestion
```

File: service/controller/v5/resource/blobobject/desired.go [hunk 4]
```diff
@@ -20,14 +22,29 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interf
...
 
+	storageAccountName := key.StorageAccountName(customObject)
 	containerName := key.BlobContainerName()
+	certificateEncryptionSecretName := key.CertificateEncryptionSecretName(customObject)
+
+	encrypter, err := r.toEncrypterObject(ctx, certificateEncryptionSecretName)
+	if apierrors.IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "encryptionkey resource is not ready")
+		resourcecanceledcontext.SetCanceled(ctx)
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
```
Comment (xh3b4sd) [inline service/controller/v5/resource/blobobject/desired.go]: [fmt](https://github.com/giantswarm/fmt/blob/master/go/log_messages.md#control-flow) says the log comes first. Below is another occasion of this.

File: service/controller/v5/templates/ignition/small.go [hunk 5]
```diff
@@ -10,6 +10,26 @@ const Small = `{
...
+        "filesystem": "root",
+        "mode": 256,
+        "contents": {
+          "source": "data:text/plain,INITIAL_VECTOR={{ .InitialVector }}"
+        }
+      }
+    ]
   }
-}
+ }
```
Comment (xh3b4sd) [inline service/controller/v5/templates/ignition/small.go]: That looks wrong.

File: service/controller/v5/resource/instance/resource.go [hunk 6]
```diff
@@ -106,3 +116,34 @@ func (r *Resource) getVMsClient(ctx context.Context) (*compute.VirtualMachineSca
 
 	return sc.AzureClientSet.VirtualMachineScaleSetVMsClient, nil
 }
+
+func (r *Resource) getEncrypterObject(ctx context.Context, secretName string) (enc.Encrypter, error) {
+	r.logger.LogCtx(ctx, "level", "debug", "message", "retrieving encryptionkey")
+
+	secret, err := r.k8sClient.CoreV1().Secrets(key.CertificateEncryptionNamespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return enc.Encrypter{}, microerror.Mask(err)
```
Comment (xh3b4sd) [inline service/controller/v5/resource/instance/resource.go]: `Encrypter` should be an interface as we want to have it replaceable. Then we can return `nil` here as zero value.

File: service/controller/v5/resource/instance/resource.go [hunk 7]
```diff
@@ -9,10 +9,14 @@ import (
 	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
 	"github.com/giantswarm/microerror"
 	"github.com/giantswarm/micrologger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
 
 	"github.com/giantswarm/azure-operator/service/controller/setting"
 	"github.com/giantswarm/azure-operator/service/controller/v5/controllercontext"
 	"github.com/giantswarm/azure-operator/service/controller/v5/debugger"
+	enc "github.com/giantswarm/azure-operator/service/controller/v5/encrypter"
```
Comment (xh3b4sd) [inline service/controller/v5/resource/instance/resource.go]: Would be nice to drop the alias.

File: service/controller/v5/cloudconfig/interface.go [hunk 8]
```diff
@@ -2,9 +2,11 @@ package cloudconfig
 
 import (
 	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/azure-operator/service/controller/v5/encrypter"
```
Comment (rossf7) [inline service/controller/v5/cloudconfig/interface.go]: Should be separate block.

File: service/controller/v5/cloudconfig/render.go [hunk 9]
```diff
@@ -3,6 +3,7 @@ package cloudconfig
 import (
 	"encoding/base64"
 
+	"github.com/giantswarm/azure-operator/service/controller/v5/encrypter"
```
Comment (rossf7) [inline service/controller/v5/cloudconfig/render.go]: Separate block here too.

**Comment by corest** [comment]
Addressed comments, added encrypter interface

**Comment by corest** [comment]
Tested manually, works fine. If no other things to add - merging. Sorry for writing so many comments. That's gonna be last PR in ignition issue :tada:

**Comment by corest** [comment]
I've tested this manually. As tests are failing even for current release, I'm merging this

**Comment by xh3b4sd** [comment]
Did you test an upgrade to this version?

**Comment by corest** [comment]
Yes, it works nice

**Comment by corest** [comment]
I mean, in general upgrade is broken in Azure. But this current upgrade works the same as upgrade between other versions

**Pull request merged**